File: exercises/exercises-mpi-p/scatterwrite.py
#!/usr/bin/env python
################################################################
################################################################
####
#### This program file is part of the book and course
####   Parallel Computing for Science and Engineering
#### by Victor Eijkhout, copyright 2013-2020
####
#### scatterwrite.py : MPI python exercise for MPI I/O
####
################################################################
################################################################

from mpi4py import MPI
import numpy as np
import struct
import time

comm = MPI.COMM_WORLD

nprocs = comm.Get_size()
procno = comm.Get_rank()

##
## Each process has a buffer of length `nwords'
## containing consecutive integers
##
nwords = 3
output_data = np.empty(nwords,dtype=np.intc)
for iw in range(nwords):
    output_data[iw] = procno*nwords + iw + 1


##
## Create the output file and open for write
##
mpifile = MPI.File.Open\
          (comm,"scatterwrite.dat",
           amode=MPI.MODE_CREATE+MPI.MODE_WRONLY)


##
## Exercise:
## -- set a file view so that each processor writes at a different offset
## -- note: offsets are in bytes
##
scattertype = MPI.INT.Create_vector(
#### your code here ####
)
scattertype.Commit()

sizeofint = np.dtype('intc').itemsize

mpifile.Set_view(
#### your code here ####
    scattertype)

nwriters = nprocs
if procno<nwriters:
    mpifile.Write( [output_data,nwords,MPI.INT] )

scattertype.Free()
mpifile.Close()

if procno==0:
    f = open("scatterwrite.dat","rb")
    location = 0
    for iw in range(nwords):
        for ip in range(nwriters):
            bytestring = f.read(sizeofint)
            ## python 3:
            #### fromfile = int.from_bytes( bytestring )
            ## python 2
            fromfile = struct.unpack("<L",bytestring)[0]

            itest = iw+1 + nwords*ip
            if fromfile!=itest:
                print("Error: @location %d, got %d, s/b %d" % \
                      (location,itest,fromfile))
            location += 1
    f.close()

if procno==0:
    print("Finished: all tasks correctly concluded.")

File: exercises/exercises-mpi-cxx/scatterwrite.cxx
/* %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   %%%%
   %%%% This program file is part of the book and course
   %%%%   "Parallel Computing for Science and Engineering"
   %%%% by Victor Eijkhout, copyright 2013-9
   %%%%
   %%%% scatterwrite.c : MPI IO example
   %%%%
   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
*/

#include <iostream>
#include <sstream>
#include <vector>
using namespace std;

#include <mpi.h>

int main(int argc,char **argv) {

  MPI_Comm comm;
  int nprocs,procno;
  stringstream proctext;

  MPI_Init(&argc,&argv);
  comm = MPI_COMM_WORLD;
  MPI_Comm_size(comm,&nprocs);
  MPI_Comm_rank(comm,&procno);

  /*
   * Each process has a buffer of length `nwords'
   * containing consecutive integers
   */
  int nwords = 3;
  vector<int> output_data(nwords);
  for (int iw=0; iw<nwords; iw++)
    output_data[iw] = procno*nwords + iw+1;
  
  MPI_File mpifile;
  MPI_File_open(comm,"scatterwrite.dat",
		MPI_MODE_CREATE | MPI_MODE_WRONLY,MPI_INFO_NULL,
		&mpifile);

  /*
   * Exercise:
   * -- set a file view so that each processor writes at a different offset
   * -- Note: offsets are in bytes
   */
  MPI_Datatype scattertype;
  MPI_Type_vector(
/**** your code here ****/
		  MPI_INT,&scattertype);
  MPI_Type_commit(&scattertype);
  MPI_Offset offset = 
/**** your code here ****/
    ;
  MPI_File_set_view
    (mpifile,
     offset,MPI_INT,scattertype,
     "native",MPI_INFO_NULL);

  int nwriters = nprocs;
  if (procno<nwriters) {
    MPI_File_write // no offset, because that is taken care of by the view
      (mpifile,output_data.data(),nwords,MPI_INT,MPI_STATUS_IGNORE);
  }

  MPI_Type_free(&scattertype);
  MPI_File_close(&mpifile);

  if (procno==0) {
    FILE *f;
    f = fopen("scatterwrite.dat","r");
    int location = 0;
    for (int iw=0; iw<nwords; iw++) { // loop over size of local buffer
      for (int ip=0; ip<nwriters; ip++) { // loop over processes
	int fromfile,success;
	success = fread(&fromfile,sizeof(int),1,f);
	if (success==EOF) {
	  proctext << "Premature end of file" << endl;
	  cerr << proctext.str(); proctext.clear();
	  break;
	}
	int itest = iw+1 + nwords*ip;
	if (fromfile!=itest) {
	  proctext << "Error " << location << ":" << itest << ":" << fromfile << endl;
	  cerr << proctext.str(); proctext.clear();
	} // else printf("Correct %d:%d:%d\n",location,itest,fromfile);
	location++;
      }
    }
    fclose(f);
    proctext << "Finished: all tasks correctly concluded." << endl;
    cerr << proctext.str(); proctext.clear();
  }

  MPI_Finalize();
  return 0;
}
